.devcontainer/launch.sh

```diff
@@ -111,83 +111,24 @@ launch_docker() {
     local -;
     set -euo pipefail
 
-    inline_vars() {
-        cat - \
-        `# inline local workspace folder` \
-      | sed "s@\${localWorkspaceFolder}@$(pwd)@g" \
-        `# inline local workspace folder basename` \
-      | sed "s@\${localWorkspaceFolderBasename}@$(basename "$(pwd)")@g" \
-        `# inline container workspace folder` \
-      | sed "s@\${containerWorkspaceFolder}@${WORKSPACE_FOLDER:-}@g" \
-        `# inline container workspace folder basename` \
-      | sed "s@\${containerWorkspaceFolderBasename}@$(basename "${WORKSPACE_FOLDER:-}")@g" \
-        `# translate local envvars to shell syntax` \
-      | sed -r 's/\$\{localEnv:([^\:]*):?(.*)\}/${\1:-\2}/g'
-    }
-
-    args_to_path() {
-        local -a keys=("${@}")
-        keys=("${keys[@]/#/[}")
-        keys=("${keys[@]/%/]}")
-        echo "$(IFS=; echo "${keys[*]}")"
-    }
-
-    json_string() {
-        python3 -c "import json,sys; print(json.load(sys.stdin)$(args_to_path "${@}"))" 2>/dev/null | inline_vars
-    }
-
-    json_array() {
-        python3 -c "import json,sys; [print(f'\"{x}\"') for x in json.load(sys.stdin)$(args_to_path "${@}")]" 2>/dev/null | inline_vars
-    }
-
-    json_map() {
-        python3 -c "import json,sys; [print(f'{k}=\"{v}\"') for k,v in json.load(sys.stdin)$(args_to_path "${@}").items()]" 2>/dev/null | inline_vars
-    }
-
-    devcontainer_metadata_json() {
-        docker inspect --type image --format '{{json .Config.Labels}}' "$DOCKER_IMAGE" \
-      | json_string '"devcontainer.metadata"'
-    }
-
     ###
     # Read relevant values from devcontainer.json
     ###
 
-    local devcontainer_json="${path}/devcontainer.json";
+    # Read and merge the devcontainer feature and `devcontainer.json` metadata
+    # Introduces the `DOCKER_IMAGE`, `ENTRYPOINTS`, `ENV_VARS`, `GPU_REQUEST`,
+    # `INITIALIZE_COMMANDS`, `MOUNTS`, `REMOTE_USER`, `RUN_ARGS`, and
+    # `WORKSPACE_FOLDER` variables
+    source <(python3 .devcontainer/launch.py "${path}/devcontainer.json")
```

Comment:
question: If we're doing this anyways, should the whole launch.sh script just be a Python script?